net: ip: net_context: AF_PACKET/SOCK_RAW/IPPROTO_RAW: set pkt family

jhirsi opened the pull request:
Setting a detected packet family (ipv4 or ipv6) in net_context level instead in lower layers for AF_PACKET/SOCK_RAW/IPPROTO_RAW type sockets when sending data.

rlubos commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -675,9 +675,24 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 				ptype = htons(NET_ETH_PTYPE_IPV6);
 				break;
-			case 0x40:
+			case 0x40: {
```
I start to wonder, if it shouldn't be `net_context` responsibility to replace `net_pkt_family()` type with `AF_INET`/`AF_INET6` instead of passing `AF_PACKET` down for `IPPROTO_RAW` sockets/contexts. For the Ethernet layer it really makes no difference what is the origin of the IP packet, and currently we're duplicating even more code.

jhirsi replied:
Excellent comment, agree, didn't like the duplication as well but couldn't found better place. 

So, you mean to set the pkt family here instead? https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/ip/net_context.c#L1766

With the quick trial it seems to work. Need to do wider testing still, also with L2/ppp. where we could also remove the similar code accordingly.

rlubos replied:
Yes, that'd be the place in the `net_context` where the packet family could be updated.

jhirsi replied:
OK, setting pkt family for IPv4 and IPv6 in net_context level and removed corresponding stuff from lower levels (ethernet and ppp). Now no additional code duplication is needed for doing ARP. Updated, PR description and subject accordingly.
Please, review.

jhirsi replied:
Unfortunately this change for this PR break the use cases of raw sckt because zephyr ip stack got involved more. Thus, fallbacking to how this PR was originally: https://github.com/zephyrproject-rtos/zephyr/pull/54664

jhirsi committed:
net: ip: net_context: AF_PACKET/SOCK_RAW/IPPROTO_RAW: set pkt family

Setting a detected packet family (ipv4 or ipv6) in net_context level
instead in lower layers for AF_PACKET/SOCK_RAW/IPPROTO_RAW type sockets
when sending data.

Signed-off-by: Jani Hirsimäki <jani.hirsimaki@nordicsemi.no>

jukkar commented on drivers/net/ppp.c:
```diff
@@ -747,21 +747,7 @@ static int ppp_send(const struct device *dev, struct net_pkt *pkt)
 			protocol = htons(PPP_IP);
 		} else if (net_pkt_family(pkt) == AF_INET6) {
 			protocol = htons(PPP_IPV6);
-		} else if (IS_ENABLED(CONFIG_NET_SOCKETS_PACKET) &&
-			   net_pkt_family(pkt) == AF_PACKET) {
-			char type = (NET_IPV6_HDR(pkt)->vtc & 0xf0);
-
-			switch (type) {
-			case 0x60:
-				protocol = htons(PPP_IPV6);
-				break;
-			case 0x40:
-				protocol = htons(PPP_IP);
-				break;
-			default:
-				return -EPROTONOSUPPORT;
-			}
-		} else {
```
Hmm, I am not sure how this is going to work. I mean that now for packet sockets, the protocol field is not set at all in this function, have you really tested that this works as expected?
Beside the `PPP_IPV6` != `NET_ETH_PTYPE_IPV6` so the protocol value would be set to incorrect value anyway.

jhirsi replied:
pkt family is now set already in net_context level and this code should be a dead code branch for 
AF_PACKET/SOCK_RAW/IPPROTO_RAW type sockets usage (to which type this code was initially added) and I have tested it to be working with that type.
But yeah, for another than IPPROTO_RAW type packets sockets, this ain't working, but it was assumed that there isn't that usage. Is that too much assumed? I can rollback this change for this file if wanted.

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -667,21 +667,6 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 			net_pkt_lladdr_src(pkt)->len =
 						sizeof(struct net_eth_addr);
 			ptype = dst_addr->sll_protocol;
-		} else if (context && net_context_get_type(context) == SOCK_RAW &&
-			   net_context_get_proto(context) == IPPROTO_RAW) {
-			char type = (NET_IPV6_HDR(pkt)->vtc & 0xf0);
-
-			switch (type) {
-			case 0x60:
-				ptype = htons(NET_ETH_PTYPE_IPV6);
-				break;
-			case 0x40:
-				ptype = htons(NET_ETH_PTYPE_IP);
-				break;
-			default:
-				ret = -ENOTSUP;
-				goto error;
-			}
```
The ptype variable is now not set for the raw sockets and that variable is given to `ethernet_fill_header()`, so to me it does not look like expected behavior.

jhirsi replied:
The idea was that a pkt family (ipv4 or ipv6) is set already in net_context level for AF_PACKET/SOCK_RAW/IPPROTO_RAW. Then later here in ethernet level when sending, an already existing code branches above are handling the ptype correctly for IPv4 and IPv6, if IS_ENABLED(CONFIG_NET_IPV4/IPV6).
This made code simpler. Otherwise, there would be a need for a duplicating similar logic from IPv4 to packet raw, see initial code in this PR.
